### PR TITLE
[SPARK-36462][K8S] Add the ability to selectively disable watching or polling

### PR DESCRIPTION
**Pull request by holdenk**
### What changes were proposed in this pull request?

Add the ability to selectively disable watching or polling

Updated version of https://github.com/apache/spark/pull/34264

### Why are the changes needed?

Watching or polling for pod status on Kubernetes can place additional load on etcd, with a large number of executors and large number of jobs this can have negative impacts and executors register themselves with the driver under normal operations anyways.

### Does this PR introduce _any_ user-facing change?

Two new config flags.

### How was this patch tested?

New unit tests + manually tested a forked version of this on an internal cluster with both watching and polling disabled.

### CODE REVIEW

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 1]
```diff
@@ -463,6 +463,25 @@ private[spark] object Config extends Logging {
...
+        "You should leave this enabled unless your encountering performance issues with your etcd.")
+      .version("3.3.0")
+      .booleanConf
+      .internal()
+      .createWithDefault(true)
+
+  val KUBERNETES_EXECUTOR_ENABLE_API_WATCHER =
+    ConfigBuilder("spark.kubernetes.executor.enableApiWatcher")
+      .doc("If Spark should create watchers for executor pod status. " +
+        "You should leave this enabled unless your encountering performance issues with your etcd.")
```
Comment (martin-g) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:478]:
```suggestion
        "You should leave this enabled unless you're encountering performance issues with your etcd.")
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsWatchSnapshotSource.scala [hunk 2]
```diff
@@ -38,19 +40,23 @@ import org.apache.spark.util.Utils
 @DeveloperApi
 class ExecutorPodsWatchSnapshotSource(
     snapshotsStore: ExecutorPodsSnapshotsStore,
-    kubernetesClient: KubernetesClient) extends Logging {
+    kubernetesClient: KubernetesClient,
+    conf: SparkConf) extends Logging {
 
   private var watchConnection: Closeable = _
+  private val enablePolling = conf.get(KUBERNETES_EXECUTOR_ENABLE_API_WATCHER)
```
Comment (martin-g) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsWatchSnapshotSource.scala:47]: `enableWatching` ?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsWatchSnapshotSource.scala [hunk 3]
```diff
@@ -38,19 +40,23 @@ import org.apache.spark.util.Utils
...
-    require(watchConnection == null, "Cannot start the watcher twice.")
-    logDebug(s"Starting watch for pods with labels $SPARK_APP_ID_LABEL=$applicationId," +
-      s" $SPARK_ROLE_LABEL=$SPARK_POD_EXECUTOR_ROLE.")
-    watchConnection = kubernetesClient.pods()
-      .withLabel(SPARK_APP_ID_LABEL, applicationId)
-      .withLabel(SPARK_ROLE_LABEL, SPARK_POD_EXECUTOR_ROLE)
-      .watch(new ExecutorPodsWatcher())
+    if (enablePolling) {
+      require(watchConnection == null, "Cannot start the watcher twice.")
+      logDebug(s"Starting watch for pods with labels $SPARK_APP_ID_LABEL=$applicationId," +
```
Comment (martin-g) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsWatchSnapshotSource.scala:53]:
```suggestion
      logDebug(s"Starting to watch for pods with labels $SPARK_APP_ID_LABEL=$applicationId," +
```

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsWatchSnapshotSourceSuite.scala [hunk 4]
```diff
@@ -61,17 +63,27 @@ class ExecutorPodsWatchSnapshotSourceSuite extends SparkFunSuite with BeforeAndA
...
+    watchSourceUnderTest.start(TEST_SPARK_APP_ID)
     val exec1 = runningExecutor(1)
     val exec2 = runningExecutor(2)
     watch.getValue.eventReceived(Action.ADDED, exec1)
     watch.getValue.eventReceived(Action.MODIFIED, exec2)
     verify(eventQueue).updatePod(exec1)
     verify(eventQueue).updatePod(exec2)
   }
+
+  test("Verify if watchers are disabled we don't call pods() on the client") {
```
Comment (martin-g) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsWatchSnapshotSourceSuite.scala:81]:
```suggestion
  test("SPARK-36462: Verify if watchers are disabled we don't call pods() on the client") {
```

**Comment by holdenk** [comment]
Thanks @martin-g for the review :)

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 5]
```diff
@@ -463,6 +463,25 @@ private[spark] object Config extends Logging {
       .checkValue(interval => interval > 0, s"Logging interval must be a positive time value.")
       .createWithDefaultString("1s")
 
+  val KUBERNETES_EXECUTOR_ENABLE_API_POLLING =
+    ConfigBuilder("spark.kubernetes.executor.enableApiPolling")
+      .doc("If Spark should poll Kubernetes for executor pod status. " +
+        "You should leave this enabled unless you're encountering performance issues with your etcd.")
+      .version("3.3.0")
```
Comment (Yikun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:470]: 3.4.0?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 6]
```diff
@@ -463,6 +463,25 @@ private[spark] object Config extends Logging {
...
+      .booleanConf
+      .internal()
+      .createWithDefault(true)
+
+  val KUBERNETES_EXECUTOR_ENABLE_API_WATCHER =
+    ConfigBuilder("spark.kubernetes.executor.enableApiWatcher")
+      .doc("If Spark should create watchers for executor pod status. " +
+        "You should leave this enabled unless you're encountering performance issues with your etcd.")
+      .version("3.3.0")
```
Comment (Yikun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:479]: same

**Review by Yikun — COMMENTED** [review]
CI failed is legit due to tiny style error. otherwise LGTM.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 7]
```diff
@@ -479,6 +479,25 @@ private[spark] object Config extends Logging {
       .checkValue(interval => interval > 0, s"Logging interval must be a positive time value.")
       .createWithDefaultString("1s")
 
+  val KUBERNETES_EXECUTOR_ENABLE_API_POLLING =
+    ConfigBuilder("spark.kubernetes.executor.enableApiPolling")
+      .doc("If Spark should poll Kubernetes for executor pod status. " +
+        "You should leave this enabled unless you're encountering issues with your etcd.")
+      .version("3.4.0")
+      .booleanConf
+      .internal()
```
Comment (dcoliversun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:488]: GA failed.
```plain
[error] /home/runner/work/spark/spark/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:488:8: value internal is not a member of org.apache.spark.internal.config.TypedConfigBuilder[Boolean]
[error] possible cause: maybe a semicolon is missing before `value internal'?
[error]       .internal()
[error]        ^
[error] /home/runner/work/spark/spark/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:497:8: value internal is not a member of org.apache.spark.internal.config.TypedConfigBuilder[Boolean]
[error] possible cause: maybe a semicolon is missing before `value internal'?
[error]       .internal()
[error]        ^
[error] two errors found
```

Comment (dcoliversun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:488]: Maybe it works.
```suggestion
      .internal()
      .booleanConf
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 8]
```diff
@@ -479,6 +479,25 @@ private[spark] object Config extends Logging {
...
+      .createWithDefault(true)
+
+  val KUBERNETES_EXECUTOR_ENABLE_API_WATCHER =
+    ConfigBuilder("spark.kubernetes.executor.enableApiWatcher")
+      .doc("If Spark should create watchers for executor pod status. " +
+        "You should leave this enabled unless you're encountering issues with your etcd.")
+      .version("3.4.0")
+      .booleanConf
+      .internal()
```
Comment (dcoliversun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:497]: ditto
```suggestion
      .internal()
      .booleanConf
```

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @holdenk . Could you address the above comments and make CIs happy?

**Comment by holdenk** [comment]
CC @dongjoon-hyun I believe done

**Comment by dongjoon-hyun** [comment]
Thank you, @holdenk .

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsWatchSnapshotSource.scala [hunk 9]
```diff
@@ -38,19 +40,23 @@ import org.apache.spark.util.Utils
 @DeveloperApi
 class ExecutorPodsWatchSnapshotSource(
     snapshotsStore: ExecutorPodsSnapshotsStore,
-    kubernetesClient: KubernetesClient) extends Logging {
+    kubernetesClient: KubernetesClient,
+    conf: SparkConf) extends Logging {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsWatchSnapshotSource.scala:44]: Oh, this is a developer API as `Stable` level. Could you make a separate constructor for this new feature?
I believe we should keep the existing 2-parameter constructer for backward compatibility.

**Review by dongjoon-hyun — COMMENTED** [review]
@holdenk . I left one comment about backward compatibility. For the other stuffs, looks good.
- https://github.com/apache/spark/pull/36433/files#r944102080

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsWatchSnapshotSource.scala (hunk 9 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsWatchSnapshotSource.scala:44]: Sure :)

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @holdenk and all.
Sorry for being late. Merged to master for Apache Spark 3.4.0.

**Pull request closed without merging**